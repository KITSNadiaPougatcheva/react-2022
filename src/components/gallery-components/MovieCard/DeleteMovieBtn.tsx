import React from "react";

export class DeleteMovieBtn extends React.PureComponent {
  constructor(public readonly props: any) { super(props); }

  render() {
    return (
      <p className="movie--delete" onClick={this.props.openModal}>Delete</p>
    );
  }
}
