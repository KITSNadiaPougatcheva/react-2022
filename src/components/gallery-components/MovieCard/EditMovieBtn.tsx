import React from "react";

export class EditMovieBtn extends React.PureComponent {
  constructor(public readonly props: any) {
    super(props);
  }

  render() {
    return (
      <p className="movie--edit" onClick={this.props.openModal}>
        Edit
      </p>
    );
  }
}
