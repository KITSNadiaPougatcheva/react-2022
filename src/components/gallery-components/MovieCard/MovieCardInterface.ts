export interface MovieCardInterface {
  details: {
    img: string | undefined;
    title: string;
    description?: string;
    short_description?: string;
    key: string;
    range: number;
    genre?: string;
  };
  key: string;
}
