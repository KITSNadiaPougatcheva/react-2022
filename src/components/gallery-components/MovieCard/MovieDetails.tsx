import React, { useCallback } from "react";

export function MovieDetails(props: any) {
  const hideDetailsModal = useCallback(() => props.hideDetails(), []);

  return (
    <>
      {props.isOpen && (
        <div className="movie-details-wrapper">
            
          <div className="movie-details-content">
            <img src={props.details.img} alt={props.details.title} />
            <div className="movie--details">
              <h2 className="movie--title">{props.details.title}</h2>
              <div className="movie--brif-description">
                {props.details.short_description}
              </div>
              <div className="movie--year">{props.details.year}</div>
              <div className="movie--range">Rating: {props.details.range}</div>
              <div className="movie--long-description">
                {props.details.description}
              </div>
              <div>
                <input
                  className="movie--close-btn"
                  type="button"
                  value="X"
                  onClick={hideDetailsModal}
                />
              </div>
            </div>
          </div>

        </div>
      )}
    </>
  );
}
