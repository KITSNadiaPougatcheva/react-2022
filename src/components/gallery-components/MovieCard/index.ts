export * from "./MovieCard";
export * from "./MovieCardInterface";

