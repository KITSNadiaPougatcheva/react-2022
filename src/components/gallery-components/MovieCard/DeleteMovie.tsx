import React from "react";
import { ModalWithButton } from "../../universal-components";
import { DeleteMovieBtn } from "./DeleteMovieBtn";

export class DeleteMovie extends React.PureComponent {
  state = {
    isOpen: false
  }

  hideModal = (event: any) => {
    event.preventDefault();
    this.setState({isOpen: false});
  }

  openModal = () => this.setState({ isOpen: true });

  render() {
    return (
      <>
        <DeleteMovieBtn openModal={this.openModal}/>
        <ModalWithButton isOpen={this.state.isOpen} hideModal={this.hideModal} title="Delete Movie"/>
      </>
    );
  }
}
