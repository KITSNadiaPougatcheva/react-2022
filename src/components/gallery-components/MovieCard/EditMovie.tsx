import { Formik } from "formik";
import React from "react";
import { connect } from "react-redux";
import { editMovieAsync } from "../../../actions";
import { ModalWithButton } from "../../universal-components";
import { EditMovieBtn } from "./EditMovieBtn";

interface Props {
  details: any;
  onEditMovie?: any;
}

class EditMovieBasic extends React.PureComponent<Props> {
  state = {
    ...this.props.details,
    isOpen: false
  };

  reset = () => {
    this.setState({ ...this.props.details, isOpen: false });
  };

  openModal = () => this.setState({ ...this.props.details, isOpen: true });

  submit = (values: any) => {
    const { title, overview, vote_average, tagline } = values;
    this.setState({ title, overview, vote_average, tagline, isOpen: false });
    const { onEditMovie } = this.props;
    onEditMovie({
      ...this.props.details,
      title,
      overview,
      vote_average: Number(vote_average),
      tagline
    });
  };

  validate = (values: any) => {
    const errors: any = {};
    if (!values.title) {
      errors.title = "Title cannot be empty";
    }
    if (!values.overview) {
      errors.overview = "Overview cannot be empty";
    }
    if (isNaN(values.vote_average)) {
      errors.vote_average = "Rating should be number";
    }
    return errors;
  };

  hideModal = (event: any) => {
    event.preventDefault();
    this.setState({ isOpen: false });
  };

  render() {
    return (
      <Formik
        initialValues={{
          title: this.props.details.title,
          overview: this.props.details.overview,
          vote_average: this.props.details.vote_average,
          tagline: this.props.details.tagline,
          openModal: this.openModal,
          hideModal: this.hideModal
        }}
        onSubmit={this.submit}
        validate={this.validate}
        onReset={this.reset}
      >
        {({
          values,
          errors,
          handleChange,
          handleSubmit,
          handleBlur,
          handleReset
        }) => (
          <>
            <EditMovieBtn openModal={values.openModal} />
            <ModalWithButton
              isOpen={this.state.isOpen}
              hideModal={handleReset}
              submit={handleSubmit}
              title="Edit Movie"
            >
              <label htmlFor="title">Title</label>
              <br />
              <input
                type="text"
                required
                name="title"
                id="title"
                role="textbox"
                value={values.title}
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.title && <div>ERR : {errors.title as string}</div>}
              <label htmlFor="overview">Description</label>
              <br />
              <input
                type="text"
                required
                name="overview"
                id="overview"
                value={values.overview}
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.overview && (
                <div>ERROR : {errors.overview as string}</div>
              )}
              <label htmlFor="tagline">Tagline</label>
              <br />
              <input
                type="text"
                required
                name="tagline"
                id="tagline"
                value={values.tagline}
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.tagline && <div>ERROR : {errors.tagline as string}</div>}
              <label htmlFor="vote_average">Rating</label>
              <br />
              <input
                type="text"
                required
                name="vote_average"
                id="vote_average"
                value={values.vote_average}
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.vote_average && (
                <div>ERROR : {errors.vote_average as string}</div>
              )}
            </ModalWithButton>
          </>
        )}
      </Formik>
    );
  }
}

const mapDispatchToProps = {
  onEditMovie: editMovieAsync
};

export const EditMovie = connect(null, mapDispatchToProps)(EditMovieBasic);
