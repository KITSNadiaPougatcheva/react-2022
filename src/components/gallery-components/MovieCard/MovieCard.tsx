import PropTypes from "prop-types";
import React from "react";
import { MovieCardInterface } from "./MovieCardInterface";

export function MovieCard(props: MovieCardInterface) {
  // throw new Error('My Test Error') // uncomment for testing puspose
  return (
    <div>
      <img src={props.img} alt={props.title} />
      <h3 className="movie--title">{props.title}</h3>
      <p className="movie--description">{props.description}</p>
    </div>
  );
}

MovieCard.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired
};
