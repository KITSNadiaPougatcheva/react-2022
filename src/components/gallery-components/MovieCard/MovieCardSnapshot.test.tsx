import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MovieCard } from "./MovieCard";

import { BrowserRouter as Router } from "react-router-dom";

const mockStore = configureMockStore([thunk]);
describe("MovieCard", () => {
  const store = mockStore({
    isOpen: true
  });

  test("renders movie tile", () => {
    const details = {
      id: "12345",
      title: "Test Movie",
      poster_path: "http://test.com/test-img.jpg",
      vote_average: 5,
      release_date: "2005/11/02"
    };

    const { asFragment } = render(
      <Router>
        <Provider store={store}>
          <MovieCard details={details} />
        </Provider>
      </Router>
    );
    expect(screen.getByText(/Edit/i)).toBeInTheDocument();
    expect(screen.getByText(/Delete/i)).toBeInTheDocument();
    expect(screen.getByAltText(details.title)).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
