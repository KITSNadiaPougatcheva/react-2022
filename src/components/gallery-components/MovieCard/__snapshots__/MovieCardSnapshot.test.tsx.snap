// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard renders movie tile 1`] = `
<DocumentFragment>
  <div
    class="movie-card-content"
  >
    <img
      alt="Test Movie"
      src="http://test.com/test-img.jpg"
    />
    <h3
      class="movie--title"
    >
      Test Movie
    </h3>
    <p
      class="movie--rating"
    >
      Rating : 5 | Release date : 2005/11/02
    </p>
    <p
      class="movie--description"
    />
    <p
      class="movie--edit"
    >
      Edit
    </p>
    <p
      class="movie--delete"
    >
      Delete
    </p>
  </div>
</DocumentFragment>
`;
